rust-runtime/aws-smithy-runtime-api/src/runtime_plugin.rs

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::config_bag::ConfigBag;
+
+type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub trait RuntimePlugin {
+    fn configure(&self, cfg: &mut ConfigBag) -> Result<(), BoxError>;
+}
+
+impl<T> From<T> for Box<dyn RuntimePlugin>
+where
+    T: RuntimePlugin + 'static,
+{
+    fn from(t: T) -> Self {
+        Box::new(t) as _
+    }
+}
+
+#[derive(Default)]
+pub struct RuntimePlugins {
+    client_plugins: Vec<Box<dyn RuntimePlugin>>,
+    operation_plugins: Vec<Box<dyn RuntimePlugin>>,
```

Comment:
`Vec` would be a good start. We may later want to change them to `VecDeque` if we need to prepend plugins.